File: testcases.py
from project import *

# PRESIDENTS

presidents_schema = DatabaseSchema("Presidents Schema")
presidents_schema.add_relation("Businessman", 1)
presidents_schema.add_relation("Economist", 1)
presidents_schema.add_relation("Democrat", 1)
presidents_schema.add_relation("Republican", 1)
presidents_schema.add_relation("Father", 2)

presidents_instance = DatabaseInstance("Presidents Instance", presidents_schema)
presidents_instance.add_fact("Businessman", ("donald", ))
presidents_instance.add_fact("Businessman", ("fred", ))
presidents_instance.add_fact("Businessman", ("james", ))
presidents_instance.add_fact("Economist", ("barack-sr", ))
presidents_instance.add_fact("Democrat", ("barack", ))
presidents_instance.add_fact("Democrat", ("franklin", ))
presidents_instance.add_fact("Republican", ("donald", ))
presidents_instance.add_fact("Father", ("barack-sr", "barack", ))
presidents_instance.add_fact("Father", ("fred", "donald", ))
presidents_instance.add_fact("Father", ("james", "franklin", ))

# CASE 1 - unary examples

presidents_examples_1 = LabeledExamples(1)
presidents_examples_1.add_positive_example(presidents_instance, ("franklin", ))
presidents_examples_1.add_positive_example(presidents_instance, ("barack", ))
presidents_examples_1.add_negative_example(presidents_instance, ("donald", ))

bq1 = FittingCQ(presidents_schema, 1)
bq1.add_relational_atom("Democrat", ("x1", ))

print("Case 1:\n")
membq = algorithm_P(presidents_instance, presidents_examples_1)
print(f"Algorithm P:\n{membq}\n")
print(f"Algorithm M:\n{algorithm_M(presidents_instance, presidents_examples_1, bq1)}\n")
print(f"Algorithm B:\n{algorithm_B(presidents_instance, presidents_examples_1)}\n")
print(f"Algorithm R:\n{algorithm_R(presidents_instance, presidents_examples_1)}\n")
print("\n")

"""
ANSWER

q = FittingCQ(presidents_schema, 1)
q.add_relational_atom("Democrat", ("x1", ))

because

("franklin") is a positive example

    ("franklin") in Democrat

("barack") is a positive example

    ("barack") in Democrat

("donald") is a negative example

    ("donald") not in Democrat

"""

# CASE 2 - unary examples

presidents_examples_2 = LabeledExamples(1)
presidents_examples_2.add_positive_example(presidents_instance, ("franklin", ))
presidents_examples_2.add_positive_example(presidents_instance, ("donald", ))
presidents_examples_2.add_negative_example(presidents_instance, ("barack", ))

bq2 = FittingCQ(presidents_schema, 1)
bq2.add_relational_atom("Father", ("x2", "x1"))
bq2.add_relational_atom("Businessman", ("x2", ))

print("Case 2:\n")
membq = algorithm_P(presidents_instance, presidents_examples_2)
print(f"Algorithm P:\n{membq}\n")
print(f"Algorithm M:\n{algorithm_M(presidents_instance, presidents_examples_2, bq2)}\n")
print(f"Algorithm B:\n{algorithm_B(presidents_instance, presidents_examples_2)}\n")
print(f"Algorithm R:\n{algorithm_R(presidents_instance, presidents_examples_2)}\n")
print("\n")

"""
ANSWER

q = FittingCQ(presidents_schema, 1)
q.add_relational_atom("Father", ("x2", "x1"))
q.add_relational_atom("Businessman", ("x2", ))

because

("franklin") is a positive example

    ("james", "franklin") in Father
    ("james") in Businessman

("donald") is a positive example

    ("fred", "donald") in Father
    ("fred") in Businessman

("barack") is a negative example

    ("barack-sr", "barack") in Father
    ("barack-sr") not in Businessman

"""

# CASE 3 - unary examples with no fitting CQ
# (should only be run against algorithm R)

presidents_examples_3 = LabeledExamples(1)
presidents_examples_3.add_positive_example(presidents_instance, ("barack", ))
presidents_examples_3.add_positive_example(presidents_instance, ("donald", ))
presidents_examples_3.add_negative_example(presidents_instance, ("franklin", ))

print("Case 3:\n")
print(f"Algorithm R:\n{algorithm_R(presidents_instance, presidents_examples_3)}\n")
print("\n")

"""
ANSWER

q = None

because

there is no fitting CQ where

    ("barack") and ("donald") are positive examples
    ("franklin") is a negative example

"""

# CASE 4 - binary examples

presidents_examples_4 = LabeledExamples(2)
presidents_examples_4.add_positive_example(presidents_instance, ("barack-sr", "barack"))
presidents_examples_4.add_negative_example(presidents_instance, ("fred", "donald"))

bq4 = FittingCQ(presidents_schema, 2)
bq4.add_relational_atom("Father", ("x1", "x2"))
bq4.add_relational_atom("Economist", ("x1", ))

print("Case 4:\n")
print(f"Algorithm P:\n{algorithm_P(presidents_instance, presidents_examples_4)}\n")
print(f"Algorithm M:\n{algorithm_M(presidents_instance, presidents_examples_4, bq4)}\n")
print(f"Algorithm B:\n{algorithm_B(presidents_instance, presidents_examples_4)}\n")
print(f"Algorithm R:\n{algorithm_R(presidents_instance, presidents_examples_4)}\n")
print("\n")

"""
ANSWER

q = FittingCQ(presidents_schema, 2)
q.add_relational_atom("Father", ("x1", "x2"))
q.add_relational_atom("Economist", ("x1", ))

because

("barack-sr", "barack") is a positive example

    ("barack-sr", "barack") in Father
    ("barack-sr") in Economist

("fred", "donald") is a negative example

    ("fred", "donald") in Father
    ("fred") not in Economist

"""

# CASE 5 - binary examples

presidents_examples_5 = LabeledExamples(2)
presidents_examples_5.add_positive_example(presidents_instance, ("fred", "donald"))
presidents_examples_5.add_negative_example(presidents_instance, ("barack-sr", "barack"))

bq5 = FittingCQ(presidents_schema, 2)
bq5.add_relational_atom("Father", ("x1", "x2"))
bq5.add_relational_atom("Republican", ("x2", ))

print("Case 5:\n")
print(f"Algorithm P:\n{algorithm_P(presidents_instance, presidents_examples_5)}\n")
print(f"Algorithm M:\n{algorithm_M(presidents_instance, presidents_examples_5, bq5)}\n")
print(f"Algorithm B:\n{algorithm_B(presidents_instance, presidents_examples_5)}\n")
print(f"Algorithm R:\n{algorithm_R(presidents_instance, presidents_examples_5)}\n")
print("\n")

"""
ANSWER

q = FittingCQ(presidents_schema, 2)
q.add_relational_atom("Father", ("x1", "x2"))
q.add_relational_atom("Republican", ("x2", ))

because

("fred", "donald") is a positive example

    ("fred", "donald") in Father
    ("donald") in Republican

("barack-sr", "barack") is a negative example

    ("barack-sr", "barack") in Father
    ("barack") not in Republican

"""

# CASE 6 - binary examples

presidents_examples_6 = LabeledExamples(2)
presidents_examples_6.add_positive_example(presidents_instance, ("james", "franklin"))
presidents_examples_6.add_negative_example(presidents_instance, ("barack-sr", "barack"))
presidents_examples_6.add_negative_example(presidents_instance, ("fred", "donald"))

bq6 = FittingCQ(presidents_schema, 2)
bq6.add_relational_atom("Businessman", ("x1", ))
bq6.add_relational_atom("Democrat", ("x2", ))

print("Case 6:\n")
print(f"Algorithm P:\n{algorithm_P(presidents_instance, presidents_examples_6)}\n")
print(f"Algorithm M:\n{algorithm_M(presidents_instance, presidents_examples_6, bq6)}\n")
print(f"Algorithm B:\n{algorithm_B(presidents_instance, presidents_examples_6)}\n")
print(f"Algorithm R:\n{algorithm_R(presidents_instance, presidents_examples_6)}\n")
print("\n")

"""
ANSWER

q = FittingCQ(presidents_schema, 2)
q.add_relational_atom("Father", ("x1", "x2"))
q.add_relational_atom("Businessman", ("x1", ))
q.add_relational_atom("Democrat", ("x2", ))

because

("james", "franklin") is a positive example

    ("james", "franklin") in Father
    ("james") in Businessman
    ("franklin") in Democrat

("barack-sr", "barack") is a negative example

    ("barack-sr", "barack") in Father
    ("barack-sr") not in Businessman
    ("barack") in Democrat

("fred", "donald") is a negative example

    ("fred", "donald") in Father
    ("fred") in Businessman
    ("donald") not in Democrat

"""

File: testcases.pl
/* PRESIDENTS */

businessman(donald).
businessman(fred).
businessman(james).
economist(baracksr).
democrat(barack).
democrat(franklin).
republican(donald).
father(baracksr, barack).
father(fred, donald).
father(james, franklin).

/* CASE 1 - unary examples */

example(1, positive, franklin).
example(1, positive, barack).
example(1, negative, donald).

/* CASE 2 - unary examples */

example(2, positive, franklin).
example(2, positive, donald).
example(2, negative, barack).

/* CASE 3 - unary examples with no fitting CQ   
   (should only be run against algorithm R)   */

example(3, positive, barack).
example(3, positive, donald).
example(3, negative, franklin).

/* CASE 4 - binary examples */

example(4, positive, baracksr, barack).
example(4, negative, fred, donald).

/* CASE 5 - binary examples */

example(5, positive, fred, donald).
example(5, negative, baracksr, barack).

/* CASE 6 - binary examples */

example(6, positive, james, franklin).
example(6, negative, baracksr, barack).
example(6, negative, fred, donald).

cq(1, positive, X1) :- example(1, positive, X1), father(X4, X2), father(X15, X7), democrat(X1), businessman(X17), businessman(X22), democrat(X2), father(X21, X13), father(X5, X3), father(X19, X12), businessman(X20), businessman(X18), businessman(X21), father(X22, X14), businessman(X16), businessman(X10), businessman(X8), father(X17, X8), businessman(X11), father(X6, X1), democrat(X14), economist(X4), democrat(X12), republican(X8), father(X18, X9).
cq(1, negative, X1) :- example(1, negative, X1).
cq(2, positive, X1) :- example(2, positive, X1), father(X15, X8), businessman(X17), businessman(X22), democrat(X2), father(X21, X13), father(X19, X12), businessman(X20), economist(X5), businessman(X18), businessman(X21), father(X22, X14), businessman(X16), father(X6, X3), businessman(X10), father(X18, X1), businessman(X9), father(X5, X2), father(X17, X9), businessman(X11), democrat(X14), democrat(X12), father(X7, X4), republican(X9), democrat(X4).
cq(2, negative, X1) :- example(2, negative, X1).
cq(4, positive, X1, X2) :- example(4, positive, X1, X2), businessman(X3), father(X6, X4), father(X5, X3), businessman(X6), father(X1, X2), republican(X3), businessman(X5), democrat(X2), democrat(X4), economist(X1).
cq(4, negative, X1, X2) :- example(4, negative, X1, X2).
cq(5, positive, X1, X2) :- example(5, positive, X1, X2), democrat(X3), businessman(X6), businessman(X1), father(X1, X2), economist(X4), democrat(X5), businessman(X2), father(X4, X3), father(X6, X5), republican(X2).
cq(5, negative, X1, X2) :- example(5, negative, X1, X2).
cq(6, positive, X1, X2) :- example(6, positive, X1, X2), democrat(X3), businessman(X6), businessman(X1), father(X1, X2), economist(X4), businessman(X5), democrat(X2), father(X4, X3), father(X6, X5), republican(X5).
cq(6, negative, X1, X2) :- example(6, negative, X1, X2).

dcq(1, positive, X1) :- distinct(cq(1, positive, X1)).
dcq(1, negative, X1) :- distinct(cq(1, negative, X1)).
dcq(2, positive, X1) :- distinct(cq(2, positive, X1)).
dcq(2, negative, X1) :- distinct(cq(2, negative, X1)).
dcq(4, positive, X1, X2) :- distinct(cq(4, positive, X1, X2)).
dcq(4, negative, X1, X2) :- distinct(cq(4, negative, X1, X2)).
dcq(5, positive, X1, X2) :- distinct(cq(5, positive, X1, X2)).
dcq(5, negative, X1, X2) :- distinct(cq(5, negative, X1, X2)).
dcq(6, positive, X1, X2) :- distinct(cq(6, positive, X1, X2)).
dcq(6, negative, X1, X2) :- distinct(cq(6, negative, X1, X2)).